File: robolectric_tests/src/com/android/launcher3/util/LauncherLayoutBuilder.java
/**
 * Copyright (C) 2019 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.android.launcher3.util;


import android.text.TextUtils;
import android.util.Pair;
import android.util.Xml;

import org.xmlpull.v1.XmlSerializer;

import java.io.IOException;
import java.io.StringWriter;
import java.io.Writer;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

/**
 * Helper class to build xml for Launcher Layout
 */
public class LauncherLayoutBuilder {

    // Object Tags
    private static final String TAG_WORKSPACE = "workspace";
    private static final String TAG_AUTO_INSTALL = "autoinstall";
    private static final String TAG_FOLDER = "folder";
    private static final String TAG_APPWIDGET = "appwidget";
    private static final String TAG_EXTRA = "extra";

    private static final String ATTR_CONTAINER = "container";
    private static final String ATTR_RANK = "rank";

    private static final String ATTR_PACKAGE_NAME = "packageName";
    private static final String ATTR_CLASS_NAME = "className";
    private static final String ATTR_TITLE = "title";
    private static final String ATTR_SCREEN = "screen";

    // x and y can be specified as negative integers, in which case -1 represents the
    // last row / column, -2 represents the second last, and so on.
    private static final String ATTR_X = "x";
    private static final String ATTR_Y = "y";
    private static final String ATTR_SPAN_X = "spanX";
    private static final String ATTR_SPAN_Y = "spanY";

    private static final String ATTR_CHILDREN = "children";


    // Style attrs -- "Extra"
    private static final String ATTR_KEY = "key";
    private static final String ATTR_VALUE = "value";

    private static final String CONTAINER_DESKTOP = "desktop";
    private static final String CONTAINER_HOTSEAT = "hotseat";

    private final ArrayList<Pair<String, HashMap<String, Object>>> mNodes = new ArrayList<>();

    public Location atHotseat(int rank) {
        Location l = new Location();
        l.items.put(ATTR_CONTAINER, CONTAINER_HOTSEAT);
        l.items.put(ATTR_RANK, Integer.toString(rank));
        return l;
    }

    public Location atWorkspace(int x, int y, int screen) {
        Location l = new Location();
        l.items.put(ATTR_CONTAINER, CONTAINER_DESKTOP);
        l.items.put(ATTR_X, Integer.toString(x));
        l.items.put(ATTR_Y, Integer.toString(y));
        l.items.put(ATTR_SCREEN, Integer.toString(screen));
        return l;
    }

    public String build() throws IOException {
        StringWriter writer = new StringWriter();
        build(writer);
        return writer.toString();
    }

    public void build(Writer writer) throws IOException {
        XmlSerializer serializer = Xml.newSerializer();
        serializer.setOutput(writer);

        serializer.startDocument("UTF-8", true);
        serializer.startTag(null, TAG_WORKSPACE);
        writeNodes(serializer, mNodes);
        serializer.endTag(null, TAG_WORKSPACE);
        serializer.endDocument();
        serializer.flush();
    }

    private static void writeNodes(XmlSerializer serializer,
            ArrayList<Pair<String, HashMap<String, Object>>> nodes) throws IOException {
        for (Pair<String, HashMap<String, Object>> node : nodes) {
            ArrayList<Pair<String, HashMap<String, Object>>> children = null;

            serializer.startTag(null, node.first);
            for (Map.Entry<String, Object> attr : node.second.entrySet()) {
                if (ATTR_CHILDREN.equals(attr.getKey())) {
                    children = (ArrayList<Pair<String, HashMap<String, Object>>>) attr.getValue();
                } else {
                    serializer.attribute(null, attr.getKey(), (String) attr.getValue());
                }
            }

            if (children != null) {
                writeNodes(serializer, children);
            }
            serializer.endTag(null, node.first);
        }
    }

    public class Location {

        final HashMap<String, Object> items = new HashMap<>();

        public LauncherLayoutBuilder putApp(String packageName, String className) {
            items.put(ATTR_PACKAGE_NAME, packageName);
            items.put(ATTR_CLASS_NAME, TextUtils.isEmpty(className) ? packageName : className);
            mNodes.add(Pair.create(TAG_AUTO_INSTALL, items));
            return LauncherLayoutBuilder.this;
        }

        public LauncherLayoutBuilder putWidget(String packageName, String className,
                int spanX, int spanY) {
            items.put(ATTR_PACKAGE_NAME, packageName);
            items.put(ATTR_CLASS_NAME, className);
            items.put(ATTR_SPAN_X, Integer.toString(spanX));
            items.put(ATTR_SPAN_Y, Integer.toString(spanY));
            mNodes.add(Pair.create(TAG_APPWIDGET, items));
            return LauncherLayoutBuilder.this;
        }

        public FolderBuilder putFolder(int titleResId) {
            FolderBuilder folderBuilder = new FolderBuilder();
            items.put(ATTR_TITLE, Integer.toString(titleResId));
            items.put(ATTR_CHILDREN, folderBuilder.mChildren);
            mNodes.add(Pair.create(TAG_FOLDER, items));
            return folderBuilder;
        }
    }

    public class FolderBuilder {

        final ArrayList<Pair<String, HashMap<String, Object>>> mChildren = new ArrayList<>();

        public FolderBuilder addApp(String packageName, String className) {
            HashMap<String, Object> items = new HashMap<>();
            items.put(ATTR_PACKAGE_NAME, packageName);
            items.put(ATTR_CLASS_NAME, TextUtils.isEmpty(className) ? packageName : className);
            mChildren.add(Pair.create(TAG_AUTO_INSTALL, items));
            return this;
        }

        public LauncherLayoutBuilder build() {
            return LauncherLayoutBuilder.this;
        }
    }
}

File: src/com/android/launcher3/util/LauncherLayoutBuilder.kt
/*
 * Copyright (C) 2019 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.launcher3.util

import android.util.Xml
import com.android.launcher3.AutoInstallsLayout.ATTR_CLASS_NAME
import com.android.launcher3.AutoInstallsLayout.ATTR_CONTAINER
import com.android.launcher3.AutoInstallsLayout.ATTR_PACKAGE_NAME
import com.android.launcher3.AutoInstallsLayout.ATTR_RANK
import com.android.launcher3.AutoInstallsLayout.ATTR_SCREEN
import com.android.launcher3.AutoInstallsLayout.ATTR_SHORTCUT_ID
import com.android.launcher3.AutoInstallsLayout.ATTR_SPAN_X
import com.android.launcher3.AutoInstallsLayout.ATTR_SPAN_Y
import com.android.launcher3.AutoInstallsLayout.ATTR_TITLE
import com.android.launcher3.AutoInstallsLayout.ATTR_TITLE_TEXT
import com.android.launcher3.AutoInstallsLayout.ATTR_USER_TYPE
import com.android.launcher3.AutoInstallsLayout.ATTR_X
import com.android.launcher3.AutoInstallsLayout.ATTR_Y
import com.android.launcher3.AutoInstallsLayout.TAG_APPWIDGET
import com.android.launcher3.AutoInstallsLayout.TAG_AUTO_INSTALL
import com.android.launcher3.AutoInstallsLayout.TAG_FOLDER
import com.android.launcher3.AutoInstallsLayout.TAG_SHORTCUT
import com.android.launcher3.AutoInstallsLayout.TAG_WORKSPACE
import com.android.launcher3.LauncherSettings.Favorites.CONTAINER_DESKTOP
import com.android.launcher3.LauncherSettings.Favorites.CONTAINER_HOTSEAT
import com.android.launcher3.LauncherSettings.Favorites.containerToString
import java.io.IOException
import java.io.StringWriter
import java.io.Writer
import org.xmlpull.v1.XmlSerializer

/** Helper class to build xml for Launcher Layout */
class LauncherLayoutBuilder {
    private val nodes = ArrayList<Node>()

    fun atHotseat(rank: Int) =
        ItemTarget(
            mapOf(
                ATTR_CONTAINER to containerToString(CONTAINER_HOTSEAT),
                ATTR_RANK to rank.toString()
            )
        )

    fun atWorkspace(x: Int, y: Int, screen: Int) =
        ItemTarget(
            mapOf(
                ATTR_CONTAINER to containerToString(CONTAINER_DESKTOP),
                ATTR_X to x.toString(),
                ATTR_Y to y.toString(),
                ATTR_SCREEN to screen.toString()
            )
        )

    @Throws(IOException::class) fun build() = StringWriter().apply { build(this) }.toString()

    @Throws(IOException::class)
    fun build(writer: Writer) {
        Xml.newSerializer().apply {
            setOutput(writer)
            setFeature("http://xmlpull.org/v1/doc/features.html#indent-output", true)
            startDocument("UTF-8", true)
            startTag(null, TAG_WORKSPACE)
            writeNodes(nodes)
            endTag(null, TAG_WORKSPACE)
            endDocument()
            flush()
        }
    }

    open inner class ItemTarget(private val baseValues: Map<String, String>) {
        @JvmOverloads
        fun putApp(packageName: String, className: String?, userType: String? = null) =
            addItem(
                TAG_AUTO_INSTALL,
                userType,
                mapOf(
                    ATTR_PACKAGE_NAME to packageName,
                    ATTR_CLASS_NAME to (className ?: packageName)
                )
            )

        @JvmOverloads
        fun putShortcut(packageName: String, shortcutId: String, userType: String? = null) =
            addItem(
                TAG_SHORTCUT,
                userType,
                mapOf(ATTR_PACKAGE_NAME to packageName, ATTR_SHORTCUT_ID to shortcutId)
            )

        @JvmOverloads
        fun putWidget(
            packageName: String,
            className: String,
            spanX: Int,
            spanY: Int,
            userType: String? = null
        ) =
            addItem(
                TAG_APPWIDGET,
                userType,
                mapOf(
                    ATTR_PACKAGE_NAME to packageName,
                    ATTR_CLASS_NAME to className,
                    ATTR_SPAN_X to spanX.toString(),
                    ATTR_SPAN_Y to spanY.toString()
                )
            )

        fun putFolder(titleResId: Int) = putFolder(ATTR_TITLE, titleResId.toString())

        fun putFolder(title: String?) = putFolder(ATTR_TITLE_TEXT, title)

        protected open fun addItem(
            tag: String,
            userType: String?,
            props: Map<String, String>,
            children: List<Node>? = null
        ): LauncherLayoutBuilder {
            nodes.add(
                Node(
                    tag,
                    HashMap(baseValues).apply {
                        putAll(props)
                        userType?.let { put(ATTR_USER_TYPE, it) }
                    },
                    children
                )
            )
            return this@LauncherLayoutBuilder
        }

        protected open fun putFolder(titleKey: String, titleValue: String?): FolderBuilder {
            val folderBuilder = FolderBuilder()
            addItem(TAG_FOLDER, null, mapOf(titleKey to (titleValue ?: "")), folderBuilder.children)
            return folderBuilder
        }
    }

    inner class FolderBuilder : ItemTarget(mapOf()) {

        val children = ArrayList<Node>()

        fun addApp(packageName: String, className: String?): FolderBuilder {
            putApp(packageName, className)
            return this
        }

        fun addShortcut(packageName: String, shortcutId: String): FolderBuilder {
            putShortcut(packageName, shortcutId)
            return this
        }

        override fun addItem(
            tag: String,
            userType: String?,
            props: Map<String, String>,
            childrenIgnored: List<Node>?
        ): LauncherLayoutBuilder {
            children.add(
                Node(tag, HashMap(props).apply { userType?.let { put(ATTR_USER_TYPE, it) } })
            )
            return this@LauncherLayoutBuilder
        }

        override fun putFolder(titleKey: String, titleValue: String?): FolderBuilder {
            throw IllegalArgumentException("Can't have folder inside a folder")
        }

        fun build() = this@LauncherLayoutBuilder
    }

    @Throws(IOException::class)
    private fun XmlSerializer.writeNodes(nodes: List<Node>) {
        nodes.forEach { node ->
            startTag(null, node.name)
            node.attrs.forEach { (key, value) -> attribute(null, key, value) }
            node.children?.let { writeNodes(it) }
            endTag(null, node.name)
        }
    }

    data class Node(
        val name: String,
        val attrs: Map<String, String>,
        val children: List<Node>? = null
    )
}
